File: backend/src/Core/Services/UrlNormalizer.cs
using System.Security.Cryptography;
using System.Text;

namespace Core.Services;

public static class UrlNormalizer
{
    private static readonly string[] TrackingParamPrefixes =
    {
        "utm_", "fbclid", "gclid", "yclid", "mc_eid", "mc_cid", "_ga", "ref", "ref_src"
    };

    private static readonly Dictionary<string, Dictionary<string, HashSet<string>>> SiteAllowlist =
        new()
        {

            ["youtube.com"] = new()
            {
                ["/watch"] = new(StringComparer.Ordinal) { "v" },
                ["/playlist"] = new(StringComparer.Ordinal) { "list" },
                ["/results"] = new(StringComparer.Ordinal) { "search_query" }
            },
            ["m.youtube.com"] = new()
            {
                ["/watch"] = new(StringComparer.Ordinal) { "v" }
            },
            ["music.youtube.com"] = new()
            {
                ["/watch"] = new(StringComparer.Ordinal) { "v" },
                ["/playlist"] = new(StringComparer.Ordinal) { "list" }
            },
            ["google.com"] = new()
            {
                ["/search"] = new(StringComparer.Ordinal) { "q" }
            },

            ["youtu.be"] = new()
            {
                ["*"] = new(StringComparer.Ordinal)
            }
        };

    private static HashSet<string>? AllowedQueryKeys(string host, string path)
    {
        if (SiteAllowlist.TryGetValue(host, out var byPath))
        {
            if (byPath.TryGetValue(path, out var keys)) return keys;
            if (byPath.TryGetValue("*", out var wildcard)) return wildcard;
        }

        if (host.EndsWith(".google.com") || host.StartsWith("google.") || host.Contains(".google."))
        {
            if (path == "/search") return new HashSet<string>(StringComparer.Ordinal) { "q" };
        }
        return null;
    }

    public static bool TryNormalize(string? input, out string normalized)
    {
        normalized = string.Empty;
        if (string.IsNullOrWhiteSpace(input)) return false;

        if (!Uri.TryCreate(input, UriKind.Absolute, out var uri)) return false;
        if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps) return false;

        var host = uri.Host.ToLowerInvariant();

        if (host.StartsWith("www.") && host[4..].Contains('.')) host = host[4..];
        var isDefault = (uri.Scheme == Uri.UriSchemeHttp && uri.Port == 80) ||
                        (uri.Scheme == Uri.UriSchemeHttps && uri.Port == 443);

        var sb = new StringBuilder();
        sb.Append(uri.Scheme).Append("://").Append(host);
        if (!isDefault) sb.Append(':').Append(uri.Port);

        var path = uri.AbsolutePath.TrimEnd('/');
        if (path.Length > 0) sb.Append(path);

        var query = uri.Query.TrimStart('?');
        if (query.Length > 0)
        {
            var allow = AllowedQueryKeys(host, path.Length == 0 ? "/" : path);
            var kept = new List<string>();
            foreach (var part in query.Split('&', StringSplitOptions.RemoveEmptyEntries))
            {
                var eq = part.IndexOf('=');
                var key = eq >= 0 ? part[..eq] : part;
                if (allow is not null)
                {
                    if (!allow.Contains(key)) continue;
                }
                else
                {
                    if (IsTrackingParam(key)) continue;
                }
                kept.Add(part);
            }
            if (kept.Count > 0)
            {
                kept.Sort(StringComparer.Ordinal);
                sb.Append('?').Append(string.Join('&', kept));
            }
        }

        normalized = sb.ToString();
        return true;
    }

    public static byte[] Hash(string normalizedUrl)
    {
        return SHA256.HashData(Encoding.UTF8.GetBytes(normalizedUrl));
    }

    private static bool IsTrackingParam(string key)
    {
        foreach (var prefix in TrackingParamPrefixes)
        {
            if (key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)) return true;
        }
        return false;
    }
}

File: extension/url_normalizer.js
(() => {
  const TRACKING_PREFIXES = [
    "utm_", "fbclid", "gclid", "yclid", "mc_eid", "mc_cid", "_ga", "ref", "ref_src"
  ];

  const SITE_ALLOWLIST = {
    "youtube.com": { "/watch": ["v"], "/playlist": ["list"], "/results": ["search_query"] },
    "m.youtube.com": { "/watch": ["v"] },
    "music.youtube.com": { "/watch": ["v"], "/playlist": ["list"] },
    "google.com": { "/search": ["q"] },
    "youtu.be": { "*": [] }
  };

  function isTracking(key) {
    const lower = key.toLowerCase();
    for (const p of TRACKING_PREFIXES) if (lower.startsWith(p)) return true;
    return false;
  }

  function allowedQueryKeys(host, path) {
    const byPath = SITE_ALLOWLIST[host];
    if (byPath) {
      if (byPath[path]) return new Set(byPath[path]);
      if (byPath["*"]) return new Set(byPath["*"]);
    }
    if (host.endsWith(".google.com") || host.startsWith("google.") || host.includes(".google.")) {
      if (path === "/search") return new Set(["q"]);
    }
    return null;
  }

  function normalizeUrl(input) {
    if (typeof input !== "string" || !input) return null;
    let u;
    try { u = new URL(input); } catch { return null; }
    if (u.protocol !== "http:" && u.protocol !== "https:") return null;

    let host = u.hostname.toLowerCase();
    if (host.startsWith("www.") && host.slice(4).includes(".")) host = host.slice(4);

    const isDefaultPort =
      (u.protocol === "http:" && (u.port === "" || u.port === "80")) ||
      (u.protocol === "https:" && (u.port === "" || u.port === "443"));

    let out = `${u.protocol}//${host}`;
    if (!isDefaultPort) out += `:${u.port}`;

    let path = u.pathname || "";
    if (path.endsWith("/") && path.length > 1) path = path.slice(0, -1);
    if (path === "/") path = "";
    out += path;

    const rawQuery = u.search.startsWith("?") ? u.search.slice(1) : u.search;
    if (rawQuery) {
      const allow = allowedQueryKeys(host, path === "" ? "/" : path);
      const kept = [];
      for (const part of rawQuery.split("&")) {
        if (!part) continue;
        const eq = part.indexOf("=");
        const key = eq >= 0 ? part.slice(0, eq) : part;
        if (allow !== null) {
          if (!allow.has(key)) continue;
        } else {
          if (isTracking(key)) continue;
        }
        kept.push(part);
      }
      if (kept.length > 0) {
        kept.sort();
        out += "?" + kept.join("&");
      }
    }

    return out;
  }

  globalThis.htNormalizeUrl = normalizeUrl;
})();
